### PR TITLE
Adding a proposal for new CRDs

### PR DESCRIPTION
**Pull request by draychev**
This adds a directory `/proposals` and a new file in it `config_ownership_crd.md`

This document will be shared with the community to gather feedback on the approach.

### CODE REVIEW

File: proposals/config_ownership_crd.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+              type: string
+            host:
+              description: "(optional) Hostname of the listener"
+              type: string
+            port:
+              description: "(required) Port number of the listener"
+              type: integer
+            path:
+              description: "(optional) URL path, for which the Ingress Controller is allowed to mutate Application Gateway configuration"
+              type: string
```
Comment (akshaysngupta) [inline proposals/config_ownership_crd.md]: how about `path` as an array of strings ?

File: proposals/config_ownership_crd.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+### Example
+A sample YAML creating new instances of this resource would have the following shape:
+```yaml
+apiVersion: "appgw.ingress.k8s.io/v1"
+kind: AzureIngressManagedListener
+metadata:
+  name: ingress-managed-listener
+spec:
+  ip: 23.45.67.89
```
Comment (akshaysngupta) [inline proposals/config_ownership_crd.md]:
```yaml
rule:
- host: prod.contoso.com
  port: 80
  path:
  - /api/*
  - /backend/*
- host: dev.contoso.com
  port: 80
  path:
  - /api/*
```

File: proposals/config_ownership_crd.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+          required:
+            - ip
+            - port
+```
+
+### Example
+A sample YAML creating new instances of this resource would have the following shape:
+```yaml
+apiVersion: "appgw.ingress.k8s.io/v1"
+kind: AzureIngressManagedListener
```
Comment (akshaysngupta) [inline proposals/config_ownership_crd.md]: How about s/`AzureIngressManagedListener`/`AzureIngressManagedLocations`

File: proposals/config_ownership_crd.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+              description: "(optional) Hostname of the listener"
+              type: string
+            port:
+              description: "(required) Port number of the listener"
+              type: integer
+            path:
+              description: "(optional) URL path, for which the Ingress Controller is allowed to mutate Application Gateway configuration"
+              type: string
+          required:
+            - ip
```
Comment (akshaysngupta) [inline proposals/config_ownership_crd.md]: how about keeping only `port` as required?

File: proposals/config_ownership_crd.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+          properties:
+            ip:
+              description: "(required) IP address of the prohibited listener; Could be the public or private address attached to the Application Gateway"
+              type: string
+            host:
+              description: "(optional) Hostname of the prohibited listener"
+              type: string
+            port:
+              description: "(required) Port number of the prohibited listener"
+              type: integer
```
Comment (akshaysngupta) [inline proposals/config_ownership_crd.md]: `port`
max: `65535`

File: proposals/config_ownership_crd.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    plural: azureingressmanagedlisteners
+  scope: Namespaced
+  validation:
+    openAPIV3Schema:
+      properties:
+        spec:
+          properties:
+            ip:
+              description: "(required) IP address of the listener managed by Ingress Controller; Could be the public or private address attached to the Application Gateway"
+              type: string
```
Comment (akshaysngupta) [inline proposals/config_ownership_crd.md]: `IPv4`/`IPv6` validation

File: proposals/config_ownership_crd.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            host:
+              description: "(optional) Hostname of the listener"
+              type: string
```
Comment (akshaysngupta) [inline proposals/config_ownership_crd.md]: `fqdn` validation

File: proposals/config_ownership_crd.md (hunk 1 above)
Comment (akshaysngupta) [reply proposals/config_ownership_crd.md]: validation that each path starts with a `/`

File: proposals/config_ownership_crd.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  - [health probes](https://github.com/Azure/application-gateway-kubernetes-ingress/blob/9820ff8aed758b2626c068e2fb25629e06540159/pkg/controller/controller.go#L76)
+
+Ingress Controller assumes full ownership and control of the given App Gateway. This makes it impossible to deploy it
+to environments with existing pre-configured App Gateways (brownfield deployments). It is also not possible to use AGIC
+in hybrid environments where Application Gateway configuration is managed in parts by system administrators as well as
+Ingress Controller.
+
+### Proposed Solution
+We propose the creation of two new Kubernetes custom resource definitions ([CRD](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)):
+  - `AzureIngressManagedListener` - defines a reference to a single listener. Ingress Controller will assume ownership
```
Comment (akshaysngupta) [inline proposals/config_ownership_crd.md]: Would it be better to have all allowed `targets/rules/locations` in one place ?

File: proposals/config_ownership_crd.md (hunk 1 above)
Comment (draychev) [reply proposals/config_ownership_crd.md]: Per our conversation - for this v1 - paths should start with `/` and end with `/*`

File: proposals/config_ownership_crd.md (hunk 6 above)
Comment (draychev) [reply proposals/config_ownership_crd.md]: I agree it'd be great to have the validators here. I'd like to keep them out of the proposal for now.
Once we iron out the details of these CRDs - when we are ready to actually commit the YAML as a separate file - we will add validations.
I'm concerned that adding a long regex will distract from the overall goal of determining whether the CRD and its shape here make sense.

File: proposals/config_ownership_crd.md (hunk 8 above)
Comment (draychev) [reply proposals/config_ownership_crd.md]: Talked offline - decided to keep one CRD per ip/port/host/[urls]

**Comment by draychev** [comment]
@akshaysngupta @asridharan `AzureIngressManagedLocation` did not sit well with me.
After working with the code for awhile `AzureIngressManagedTarget` (`...Target`) started sounding better.
We have `ingress`, flow of traffic, and it lands, arrives at a specific location. So we configure that path. I think `Target` for the ingress makes sense.

Are you ok with this rename?

**Pull request merged**